### ISSUE TITLE
'i2scSetMSArrayItem' was not declared in this scope

### ISSUE THREAD

**Issue by marcx33** [roberttidey/espI2sClockGen#4]
I'm getting error:

'i2scSetMSArrayItem' was not declared in this scope; did you mean 'setMSArrayItem'?

All required libraries are installed.

What could be the issue ?

**Comment by roberttidey** [comment]
i2scSetMSArrayItem is declared in library i2sTXcircular

setMSArrayItem is a wrapper function in the top level ino

Are you sure this library is installed in a library accessible path?

**Comment by marcx33** [comment]
Yes, it is accessible. I've tried initialising the I2S in a separate sketch and library works fine.

One more question for you, is it possible to use  your javascript calculation logic to build arduino sketch that will perform the same calculation but without javascript ? Input parameter is target clock freq. Need outputs: bit clock, mark bits, space bits, buffer size, div1, div2 (I assume this are required to generate correctly). I'd like to generate the target frequency directly from sketch without external files or webpage interface.

**Comment by roberttidey** [comment]
I have just compiled  espi2sClockgen without issues. If you still have problems with compiling this sketch with the library then try turning on verbose output during compile (file/preferences). Also what version of Arduino and esp9266 plugin are you using?

It would be possible to perform the calculations on the esp8266. The 3 key functions required are calculateDutyCycle(),  calculateClock(), and setClock().

The setClock function performs the final calculation and does the url call with the parameters to set the generator going.

It is fairly straightforward to translate the javascript to the c code for the esp8266  arduino as the basic syntax is very similar.

How do you plan to control it? Where do the input parameters come from and how will you communicate these to the esp8266?

**Comment by marcx33** [comment]
Bellow is the output of compiler. I'm using arduino 1.8.18 and D1Mini v3.

[arduino-output1.txt](https://github.com/roberttidey/espI2sClockGen/files/14085250/arduino-output1.txt)

I plan to add a membrane keypad to set the desired frequency and oled for UI.

**Comment by marcx33** [comment]
I've managed to translate calculateClock() function.  I'm having trouble translating  calculateDutyCycle(). For some frequencies calculation is the same as javascript, for some it is not.  Could you take a look:

void calculateDutyCycle(long tB) {
  long i;

  i = tB / 2; // Since the duty cycle is 50%, we divide tB by 2
  if (i == 0) i = 1; // Ensure at least one bit is allocated for marking

  Serial.print("markbits: ");
  Serial.println(i);
  Serial.print("spacebits: ");
  Serial.println(tB - i);
}

**Comment by roberttidey** [comment]
On the compile issue it looks like these may be triggered by errors being detected during processing of the BaseSupport library. I don't get those. Can you check the BaseSupport Library is OK. One way to check that independently is just to do a verify compile of the BaseSupport example.

Could the duty cycle calculation difference be due to the javascript rounding the division whereas the c code is rounding down? There will be a difference for cases where tB is an odd value. If not then please give an example of where there is a difference and the tB value used for that case.

**Comment by marcx33** [comment]
Thank you for your asistance, I can't figure what is the issue. I've managed to translate javascript code so now I can generate frequency directly from sketch.  I'm posting it here, maybe someone will need it.

After running the calculation function,  frequency is generated with this code:

 i2scInit(count32, div1 , div2, 0, 0);  
 i2scSetMSArrayItem(0, markbits, spacebits,1);
 i2scBegin(); 

[void_calculate_clock.txt](https://github.com/roberttidey/espI2sClockGen/files/14089737/void_calculate_clock.txt)
](url)
